csrc/id_graphs.cpp

```diff
@@ -0,0 +1,3620 @@
+    auto loop_covered_exact_groups =
+        idGraph(IdMappingMode::EXACT).toGroups(*loop_group);
+
+    // The intersection of the exact groups that the broadcast domains can be
+    // broadcasted to, and those that exist within the same loop groop are is
+    // the promotion needed for this iel_group.
+    auto loop_exact_resolved_intersection =
+        resolved_exact_groups.intersect(loop_covered_exact_groups);
+
+    if (loop_exact_resolved_intersection.empty()) {
+      // No resolution
+      continue;
+    }
+
+    if (loop_exact_resolved_intersection.size() > 1) {
+      std::stringstream err_msg;
+
+      err_msg
+          << "Invalid multiple broadcast resolution within shared loops detected, group:\n  "
+          << iel_group->toString() << "\nIs being broadcasted to:";
+
+      for (auto entry : loop_exact_resolved_intersection) {
+        err_msg << "\n  " << entry->toString();
+      }
+      TORCH_INTERNAL_ASSERT(false, err_msg.str());
+    }
+
+    // loop_exact_resolved_intersection.size() must be 1 at this point
+    auto exact_resolution_group = loop_exact_resolved_intersection.front();
+
+    VectorOfUniqueEntries<IterDomain*> resolved_ids =
+        exact_resolution_group->intersect(*loop_group);
+    auto promoted_iel_groups =
+        intersection_exact_loop_graph.toGroups(resolved_ids);
+
+    if (promoted_iel_groups.size() == 0) {
+      continue;
+    }
+
+    if (promoted_iel_groups.size() > 1) {
+      std::stringstream err_msg;
+
+      err_msg
+          << "Invalid multiple broadcast resolution within shared loops detected, group:\n  "
+          << iel_group->toString() << "\nIs being broadcasted to:";
+
+      for (auto entry : promoted_iel_groups) {
+        err_msg << "\n  " << entry->toString();
+      }
+      TORCH_INTERNAL_ASSERT(false, err_msg.str());
+    }
+
+    iel_promotion_map[iel_group] = promoted_iel_groups.front()->front();
+  }
+
+  std::cout << "Initial promotion map:" << std::endl;
+
+  for (auto iel_group :
+       intersection_exact_loop_graph.disjointIdSets().disjointSets()) {
+    auto entry_it = iel_promotion_map.find(iel_group);
+    if (entry_it == iel_promotion_map.end()) {
+      continue;
+    }
+    std::cout << "  " << entry_it->second->toString() << " <- "
+              << entry_it->first->toString() << std::endl;
+  }
+
+  IdGraphStmtSort iel_stmt_sort(intersection_exact_loop_graph);
+
+  std::cout << "Initial promotion replay:" << std::endl;
+  for (auto iel_expr : iel_stmt_sort.exprs()) {
+    IdGroups input_groups = intersection_exact_loop_graph.inputGroups(iel_expr);
+
+    // Check if any inputs need promotion indicating this expr group needs to
+    // be replayed with promoted inputs
+    std::vector<IterDomain*> promoted_inputs;
+    bool an_input_was_promoted = false;
+
+    for (auto inp : input_groups) {
+      auto inp_promo_it = iel_promotion_map.find(inp);
+      if (inp_promo_it == iel_promotion_map.end()) {
+        promoted_inputs.push_back(inp->front());
```

Comment:
Is it equivalent to save `inp` in `non_promoted_input_uses` here instead of `promoted_input_groups.intersect(input_groups)` at lines around 2554?

Reply:
Yeah, I think so.

Reply:
Stale.